README.md
[![NPM version](https://badge.fury.io/js/cdk-fargate-patterns.svg)](https://badge.fury.io/js/cdk-fargate-patterns)
[![PyPI version](https://badge.fury.io/py/cdk-fargate-patterns.svg)](https://badge.fury.io/py/cdk-fargate-patterns)
[![Release](https://github.com/pahud/cdk-fargate-patterns/actions/workflows/release.yml/badge.svg)](https://github.com/pahud/cdk-fargate-patterns/actions/workflows/release.yml)
<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-6-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->

# cdk-fargate-patterns

CDK patterns for serverless container with AWS Fargate

# `DualAlbFargateService`


![](images/DualAlbFargateService.svg)

Inspired by _Vijay Menon_ from the [AWS blog post](https://aws.amazon.com/blogs/containers/how-to-use-multiple-load-balancer-target-group-support-for-amazon-ecs-to-access-internal-and-external-service-endpoint-using-the-same-dns-name/) introduced in 2019, `DualAlbFargateService` allows you to create one or many fargate services with both internet-facing ALB and internal ALB associated with all services. With this pattern, fargate services will be allowed to intercommunicat via internal ALB while external inbound traffic will be spread across the same service tasks through internet-facing ALB.


The sample below will create 3 fargate services associated with both external and internal ALBs. The internal ALB will have an alias(`internal.svc.local`) auto-configured from Route 53 so services can interconnect through the private ALB endpoint.


```ts
new DualAlbFargateService(stack, 'Service', {
  spot: true, // FARGATE_SPOT only cluster
  tasks: [
    {
      task: orderTask,
      desiredCount: 2,
      external: { port: 443, certificate },
      internal: { port: 80 },
      // customize the service autoscaling policy
      scalingPolicy: {
        maxCapacity: 20,
        requestPerTarget: 1000,
        targetCpuUtilization: 50,
      },
    },
    { task: customerTask, desiredCount: 2, internal: { port: 8080 } },
    { task: productTask, desiredCount: 2, internal: { port: 9090 } },
    
    // support gRPC application.
    // you can use protocolVersion to setting gRPC for your application. 
    // you need to specify acm and grpc port.
    // and don't forget to setting healthCheck.healthyGrpcCodes.  
    {
        task: task,
        desiredCount: 1,
        internal: { port: 50051, certificate: [cert] },
        external: { port: 50051, certificate: [cert] },
        protocolVersion: elbv2.ApplicationProtocolVersion.GRPC,
        healthCheck: {
          healthyGrpcCodes: '12',
        },
      },
  ],
  route53Ops: {
    zoneName: 'svc.local',
    externalElbRecordName: 'external',
    internalElbRecordName: 'internal',
  },
});
```

# `DualNlbFargateService`

Similar to `DualAlbFargateService`, you are allowed to deploy multiple container services with AWS Fargate as well as external NLB and internal NLB.

To allowed ingress traffic, you will need to explicitly add ingress rules on the `connections`:

```ts
const nlbService = new DualNlbFargateService(stack, 'NLBService', {
  tasks: [...],
});

// we need this to allow ingress traffic from public internet only for the order service
nlbService.service[0].connections.allowFromAnyIpv4(ec2.Port.tcp(8080));
// allow from VPC
nlbService.service.forEach(s => {
  s.connections.allowFrom(ec2.Peer.ipv4(nlbService.vpc.vpcCidrBlock), ec2.Port.tcp(8080));
});

```


## Fargate Spot Support

By enabling the `spot` property, 100% fargate spot tasks will be provisioned to help you save up to 70%. Check more details about [Fargate Spot](https://aws.amazon.com/about-aws/whats-new/2019/12/aws-launches-fargate-spot-save-up-to-70-for-fault-tolerant-applications/?nc1=h_ls). This is a handy catch-all flag to force all tasks to be `FARGATE_SPOT` only.

To specify mixed strategy with partial `FARGATE` and partial `FARGATE_SPOT`, specify the `capacityProviderStrategy` for individual tasks like.


```ts
new DualAlbFargateService(stack, 'Service', {
  tasks: [
    {
      task: customerTask,
      internal: { port: 8080 },
      desiredCount: 2,
      capacityProviderStrategy: [
        {
          capacityProvider: 'FARGATE',
          base: 1,
          weight: 1,
        },
        {
          capacityProvider: 'FARGATE_SPOT',
          base: 0,
          weight: 3,
        },
      ],
    },
  ],
});
```
The custom capacity provider strategy will be applied if `capacityProviderStretegy` is specified, otherwise, 100% spot will be used when `spot: true`. The default policy is 100% Fargate on-demand.

### Fargate Spot Termination Handling

By default, if fargate spot capacity is available in the cluster, a fargate spot termination handler Lambda function will be created with proper IAM role policies to handle the termination event to ensure we deregister the fargate spot task from target groups gracefully. While it's a recommended feature, you may opt out with `spotTerminationHandler: false`.

```ts
new DualAlbFargateService(stack, 'Service', {
  spot: true, // FARGATE_SPOT only cluster
  spotTerminationHandler: false,  // opt-out
  ...
}
```

## ECS Exec

Simply turn on the `enableExecuteCommand` property to enable the [ECS Exec](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-exec.html) support for all services.

## ECS deployment circuit breaker

ECS deployment circuit breaker automatically rolls back unhealthy service deployments without the need for manual intervention. By default this feature is enabled, you can opt out with `circuitBreaker: false`. Read the [docummentation](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/deployment-type-ecs.html) or [blog post](https://aws.amazon.com/tw/blogs/containers/announcing-amazon-ecs-deployment-circuit-breaker/) for more details.

## Internal, External or Both

Specify the `internal` or `external` property to expose your service internally, externally or both.

The `certificate` property implies `HTTPS` protocol.

```ts
new DualAlbFargateService(stack, 'Service', {
    tasks: [
      // this task is internal only
      { task: task1, internal: { port: 8080 } },
      // this task is external only
      { task: task2, external: { port: 8081 } },
      // this task is both external(HTTPS) and internal(HTTP) facing
      { 
        task: task3,
        external: { port: 443, certificate: myAcmCert },
        internal: { port: 8888 },
      },
    ],
  });
```

## VPC Subnets

By default, all tasks will be deployed in the private subnets. You will need the NAT gateway for the default route associated with the private subnets to ensure the task can successfully pull the container images.

However, you are allowed to specify `vpcSubnets` to customize the subnet selection.

To deploy all tasks in public subnets, one per AZ:

```ts
new DualAlbFargateService(stack, 'Service', {
    vpcSubnets: {
      subnetType: ec2.SubnetType.PUBLIC,
      onePerAz: true,
  },
  ...
});
```
This will implicitly enable the `auto assign public IP` for each fargate task so the task can successfully pull the container images from external registry. However, the ingress traffic will still be balanced via the external ALB.

To deploy all tasks in specific subnets:

```ts
new DualAlbFargateService(stack, 'Service', {
  vpcSubnets: { 
      subnets: [
        ec2.Subnet.fromSubnetId(stack, 'sub-1a', 'subnet-0e9460dbcfc4cf6ee'),
        ec2.Subnet.fromSubnetId(stack, 'sub-1b', 'subnet-0562f666bdf5c29af'),
        ec2.Subnet.fromSubnetId(stack, 'sub-1c', 'subnet-00ab15c0022872f06'),
      ],
    },
  ...
});
```

### Import an existing Cluster
!!! Before using an existing `ECS Cluster`, please make sure you have the following:
- see : [Adding Fargate capacity providers to an existing cluster](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/fargate-capacity-providers.html#fargate-capacity-providers-existing-cluster)
```ts
const vpc = ec2.Vpc.fromLookup(stack, 'defVpc', { isDefault: true });
const sg = new ec2.SecurityGroup(stack, 'demo-sg', {
  vpc,
});
const existCluster = ecs.Cluster.fromClusterAttributes(stack, 'existCluster', {
  securityGroups: [sg],
  clusterName: 'existCluster',
  vpc,
});

new DualAlbFargateService(stack, 'Service', {
  enableExecuteCommand: true,
  cluster: existCluster,
  tasks: [
    {
      task: nginx,
      desiredCount: 1,
      external: { port: 80 },
      capacityProviderStrategy: [{
        capacityProvider: 'FARGATE_SPOT',
        weight: 1,
      }],
    },
  ],
});
```


## Sample Application

This repository comes with a sample applicaiton with 3 services in Golang. On deployment, the `Order` service will be exposed externally on external ALB port `80` and all requests to the `Order` service will trigger sub-requests internally to another other two services(`product` and `customer`) through the internal ALB and eventually aggregate the response back to the client.


![](images/DualAlbFargateService.svg)

## Deploy

To deploy the sample application in you default VPC:

```sh
// install first
$ yarn install
// compile the ts to js
$ yarn build
$ npx cdk --app lib/integ.default.js -c use_default_vpc=1 diff
$ npx cdk --app lib/integ.default.js -c use_default_vpc=1 deploy
```

To deploy with HTTPS-only with existing ACM certificate in your default VPC:

```sh
$ npx cdk --app lib/integ.default.js deploy -c use_default_vpc=1 -c ACM_CERT_ARN=<YOUR_ACM_CERT_ARN>
```

On deployment complete, you will see the external ALB endpoint in the CDK output. `cURL` the external HTTP endpoint and you should be able to see the aggregated response.


```sh
$ curl http://demo-Servi-EH1OINYDWDU9-1397122594.ap-northeast-1.elb.amazonaws.com
or
$ curl https://<YOUR_CUSTOM_DOMAIN_NAME>

{"service":"order", "version":"1.0"}
{"service":"product","version":"1.0"}
{"service":"customer","version":"1.0"}
```

# `WordPress`

Use the `WordPress` construct to create a serverless **WordPress** service with AWS Fargate, Amazon EFS filesystem and Aurora serverless database. All credentials auto generated from the **AWS Secret Manager** and securely inject the credentials into the serverless container with the auto generated IAM task execution role.

```ts
new WordPress(stack, 'WP', {
  auroraServerless: true,
  spot: true,
  enableExecuteCommand: true,
});
```

# `Laravel`

The `Laravl` construct is provided as a high-level abstraction that allows you to create a modern Laravel environment running on `AWS Fargate` and `Amazon Aurora Serverless`. Here comes two variants as the reference:

**laravel-bitnami** - based on [bitnami/laravel](https://hub.docker.com/r/bitnami/laravel/) with `artisan` running as the built-in web server.

**laravel-nginx-php-fpm** - laravel with nginx and php-fpm maintained by [Ernest Chiang](https://github.com/dwchiang).

Simply point `code` to your local Laravel codebase and it takes care everything else for you.

## Samples

```ts
/**
 * laravel-bitnami
 */
new Laravel(stack, 'LaravelBitnamiDemo', {
  auroraServerless: true,
  spot: true,
  enableExecuteCommand: true,
  code: path.join(__dirname, '../services/laravel-bitnami'),
  containerPort: 3000,
  loadbalancer: { port: 80 },
});

/**
 * laravel-nginx-php-fpm
 */
new Laravel(stack, 'LaravelNginxDemo', {
  auroraServerless: true,
  spot: true,
  enableExecuteCommand: true,
  code: path.join(__dirname, '../services/laravel-nginx-php-fpm'),
  loadbalancer: { port: 80 },
});
```

See [integ.laravel.ts](src/integ.laravel.ts) for the full code sample.


# Local development and testing

The [docker-compose.yml](./services/docker-compose.yml) is provided with all sample services in the repository. To bring up all services locally, run:

```sh
$ cd services
$ docker compose up
```
Use `cURL` to test locally:

```
curl http://localhost
```
Response:
```
{"service":"order","version":"1.0"}
{"service":"product","version":"1.0"}
{"service":"customer","version":"1.0"}
```

# FAQ

Q: What is the difference between `cdk-fargate-patterns` and [aws-ecs-patterns](https://github.com/aws/aws-cdk/tree/master/packages/%40aws-cdk/aws-ecs-patterns)?

A: `aws-ecs-patterns` comes with a few patterns around Amazon ECS with AWS Fargate or AWS EC2 and focuses on some scenarios with single service and single ELB like `ApplicationLoadBalancedFargateService` and `NetworkLoadBalancedFargateService`. However, `cdk-fargate-patterns` is trying to explore use cases on modern application which usually comes with multiple container services grouped as a deployment with inter-service connectivity as well as ingress traffic from external internet by seperating the internal ELB from the external one.

## Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://blog.neilkuan.net"><img src="https://avatars.githubusercontent.com/u/46012524?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Neil Kuan</b></sub></a><br /><a href="#design-neilkuan" title="Design">🎨</a> <a href="https://github.com/pahud/cdk-fargate-patterns/commits?author=neilkuan" title="Code">💻</a> <a href="https://github.com/pahud/cdk-fargate-patterns/commits?author=neilkuan" title="Tests">⚠️</a> <a href="#example-neilkuan" title="Examples">💡</a></td>
    <td align="center"><a href="https://www.ernestchiang.com"><img src="https://avatars.githubusercontent.com/u/251263?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Ernest Chiang</b></sub></a><br /><a href="#ideas-dwchiang" title="Ideas, Planning, & Feedback">🤔</a> <a href="https://github.com/pahud/cdk-fargate-patterns/commits?author=dwchiang" title="Tests">⚠️</a> <a href="https://github.com/pahud/cdk-fargate-patterns/commits?author=dwchiang" title="Code">💻</a> <a href="#example-dwchiang" title="Examples">💡</a></td>
    <td align="center"><a href="https://clarence.tw"><img src="https://avatars.githubusercontent.com/u/5698291?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Clarence</b></sub></a><br /><a href="#example-clarencetw" title="Examples">💡</a> <a href="https://github.com/pahud/cdk-fargate-patterns/commits?author=clarencetw" title="Code">💻</a></td>
    <td align="center"><a href="https://github.com/paper5487"><img src="https://avatars.githubusercontent.com/u/15643225?v=4?s=100" width="100px;" alt=""/><br /><sub><b>paper5487</b></sub></a><br /><a href="https://github.com/pahud/cdk-fargate-patterns/issues?q=author%3Apaper5487" title="Bug reports">🐛</a></td>
    <td align="center"><a href="https://github.com/plarsson"><img src="https://avatars.githubusercontent.com/u/903607?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Peter Larsson</b></sub></a><br /><a href="#ideas-plarsson" title="Ideas, Planning, & Feedback">🤔</a></td>
    <td align="center"><a href="https://github.com/LeoChien-SC"><img src="https://avatars.githubusercontent.com/u/81736089?v=4?s=100" width="100px;" alt=""/><br /><sub><b>LeoChien-SC</b></sub></a><br /><a href="https://github.com/pahud/cdk-fargate-patterns/issues?q=author%3ALeoChien-SC" title="Bug reports">🐛</a></td>
  </tr>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.3.1](https://github.com/pahud/cdk-fargate-patterns/compare/v0.3.0...v0.3.1) (2021-07-24)

## [0.2.0](https://github.com/pahud/cdk-fargate-patterns/compare/v0.1.8...v0.2.0) (2021-07-06)


src/integ.default.ts
import * as path from 'path';
import * as acm from '@aws-cdk/aws-certificatemanager';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecs from '@aws-cdk/aws-ecs';
import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';
import * as cdk from '@aws-cdk/core';
import { DualAlbFargateService } from './index';

const app = new cdk.App();

const env = {
  region: process.env.CDK_DEFAULT_REGION,
  account: process.env.CDK_DEFAULT_ACCOUNT,
};

const stack = new cdk.Stack(app, 'demo-stack2', { env });

const zoneName = 'svc.local';
const internalAlbRecordName = 'internal';
const externalAlbRecordName = 'external';
const internalALBEndpoint = `http://${internalAlbRecordName}.${zoneName}`;

// order service
const orderTask = new ecs.FargateTaskDefinition(stack, 'orderTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

orderTask.addContainer('order', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/golang/OrderService')),
  portMappings: [
    { containerPort: 8080 },
  ],
  environment: {
    PRODUCT_SVC_URL: `${internalALBEndpoint}:9090`,
    CUSTOMER_SVC_URL: `${internalALBEndpoint}:8080`,
    WAREHOUSE_SVC_URL: `${internalALBEndpoint}:9092/api.php`,
    serviceName: 'order',
    versionNum: '1.0',
  },
});

// customer service
const customerTask = new ecs.FargateTaskDefinition(stack, 'customerTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

customerTask.addContainer('customer', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/golang/CommonService')),
  portMappings: [
    { containerPort: 8080 },
  ],
  environment: {
    PRODUCT_SVC_URL: `${internalALBEndpoint}:9090`,
    CUSTOMER_SVC_URL: `${internalALBEndpoint}:8080`,
    serviceName: 'customer',
    versionNum: '1.0',
  },
});

// product service
const productTask = new ecs.FargateTaskDefinition(stack, 'productTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

productTask.addContainer('product', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/golang/CommonService')),
  portMappings: [
    { containerPort: 8080 },
  ],
  environment: {
    PRODUCT_SVC_URL: `${internalALBEndpoint}:9090`,
    CUSTOMER_SVC_URL: `${internalALBEndpoint}:8080`,
    serviceName: 'product',
    versionNum: '1.0',
  },
});

// nginx service
const nginxTask = new ecs.FargateTaskDefinition(stack, 'nginxTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

nginxTask.addContainer('nginx', {
  image: ecs.ContainerImage.fromRegistry('nginx:latest'),
  portMappings: [
    { containerPort: 80 },
  ],
});

// php service
const phpTask = new ecs.FargateTaskDefinition(stack, 'phpTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

phpTask.addContainer('php', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/nginx-php')),
  portMappings: [
    { containerPort: 80 },
  ],
  environment: {
    serviceName: 'warehouse',
    versionNum: '0.1',
  },
});

// laravel-nginx-php-fpm service
const laravelNginxPhpFpmTask = new ecs.FargateTaskDefinition(stack, 'laravelNginxPhpFpmTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

laravelNginxPhpFpmTask.addContainer('laravel-nginx-php-fpm', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/laravel-nginx-php-fpm')),
  portMappings: [
    { containerPort: 80 },
  ],
});

// laravel-bitnami service
const laravelBitnamiTask = new ecs.FargateTaskDefinition(stack, 'laravelBitnamiTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

laravelBitnamiTask.addContainer('laravel-bitnami', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/laravel-bitnami')),
  portMappings: [
    { containerPort: 3000 },
  ],
});

// NuxtJS service
const nuxtTask = new ecs.FargateTaskDefinition(stack, 'nuxtTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

nuxtTask.addContainer('nuxt', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/nuxt')),
  portMappings: [
    { containerPort: 80 },
  ],
});

// Node.js service
const nodeTask = new ecs.FargateTaskDefinition(stack, 'nodeTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

nodeTask.addContainer('node', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/node')),
  portMappings: [
    { containerPort: 80 },
  ],
});

// java spring boot service
const javaTask = new ecs.FargateTaskDefinition(stack, 'javaTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

javaTask.addContainer('java', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/java-spring-boot')),
  portMappings: [
    { containerPort: 8080 },
  ],
});

// apache-php service
const apachePHP = new ecs.FargateTaskDefinition(stack, 'apachephpTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

apachePHP.addContainer('apachephp', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/apache-php')),
  portMappings: [
    { containerPort: 80 },
  ],
});

// python-flask service
const flaskTask = new ecs.FargateTaskDefinition(stack, 'flaskTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

flaskTask.addContainer('flask', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/python-flask')),
  portMappings: [
    { containerPort: 80 },
  ],
});

// gorilla-mux
const muxTask = new ecs.FargateTaskDefinition(stack, 'muxTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

muxTask.addContainer('mux', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/golang/gorilla-mux')),
  portMappings: [
    { containerPort: 8080 },
  ],
});

// go-grpc
const gogrpc = new ecs.FargateTaskDefinition(stack, 'gogrpcTask', {
  cpu: 256,
  memoryLimitMiB: 512,
});

gogrpc.addContainer('gogrpc', {
  image: ecs.ContainerImage.fromAsset(path.join(__dirname, '../services/go-grpc')),
  portMappings: [
    { containerPort: 50051 },
  ],
});

const certArn = stack.node.tryGetContext('ACM_CERT_ARN');
const cert = certArn ? acm.Certificate.fromCertificateArn(stack, 'Cert', certArn) : undefined;

const svc = new DualAlbFargateService(stack, 'Service', {
  spot: true, // FARGATE_SPOT only cluster
  enableExecuteCommand: true,
  tasks: [
    // The order service with both external/internal access
    {
      task: orderTask,
      desiredCount: 2,
      internal: { port: 80 },
      external: cert ? { port: 443, certificate: [cert] } : { port: 80 },
      // customize the service autoscaling policy
      scalingPolicy: {
        maxCapacity: 20,
        requestPerTarget: 1000,
        targetCpuUtilization: 50,
      },
    },
    {
      // The customer service(internal only)
      task: customerTask,
      desiredCount: 1,
      internal: { port: 8080 },
      capacityProviderStrategy: [
        {
          capacityProvider: 'FARGATE',
          base: 1,
          weight: 1,
        },
        {
          capacityProvider: 'FARGATE_SPOT',
          base: 0,
          weight: 3,
        },
      ],
    },
    // The produce service(internal only)
    {
      task: productTask,
      desiredCount: 1,
      internal: { port: 9090 },
    },
    // The nginx service(external only)
    {
      task: nginxTask,
      desiredCount: 1,
      external: { port: 9091 },
    },
    // The nginx-php-fpm service(external/internal)
    {
      task: phpTask,
      desiredCount: 1,
      internal: { port: 9092 },
      external: { port: 9092 },
    },
    // The NuxtJS service(external/internal)
    {
      task: nuxtTask,
      desiredCount: 1,
      internal: { port: 9093 },
      external: { port: 9093 },
    },
    // The node service(external/internal)
    {
      task: nodeTask,
      desiredCount: 1,
      internal: { port: 9094 },
      external: { port: 9094 },
    },
    // The laravel-nginx-php-fpm service(external/internal)
    {
      task: laravelNginxPhpFpmTask,
      desiredCount: 1,
      internal: { port: 9095 },
      external: { port: 9095 },
    },
    // The laravel-bitnami service(external/internal)
    {
      task: laravelBitnamiTask,
      desiredCount: 1,
      internal: { port: 9096 },
      external: { port: 9096 },
    },
    // java spring boot service(external/internal)
    {
      task: javaTask,
      desiredCount: 1,
      internal: { port: 9097 },
      external: { port: 9097 },
      healthCheck: { path: '/hello-world' },
    },
    // apache-php(external/internal)
    {
      task: apachePHP,
      desiredCount: 1,
      internal: { port: 9098 },
      external: { port: 9098 },
    },
    // python-flask(external/internal)
    {
      task: flaskTask,
      desiredCount: 1,
      internal: { port: 9099 },
      external: { port: 9099 },
    },
    // gorilla-mux(external/internal)
    {
      task: muxTask,
      desiredCount: 1,
      internal: { port: 9100 },
      external: { port: 9100 },
    },
    // go-grpc(external/internal)
    {
      task: gogrpc,
      desiredCount: 1,
      internal: cert ? { port: 50051, certificate: [cert] } : { port: 80 },
      external: cert ? { port: 50051, certificate: [cert] } : { port: 80 },
      protocolVersion: elbv2.ApplicationProtocolVersion.GRPC,
      healthCheck: {
        healthyGrpcCodes: '12',
      },
    },
  ],
  route53Ops: {
    zoneName, // svc.local
    externalElbRecordName: externalAlbRecordName, // external.svc.local
    internalElbRecordName: internalAlbRecordName, // internal.svc.local
  },
});


// create a dummy sg
const dummySg = new ec2.SecurityGroup(stack, 'DummySG', {
  vpc: svc.vpc,
});

// allow all traffic from dummy sg to all the services
for (let i = 0; i < svc.service.length; i++) {
  svc.service[i].connections.allowFrom(dummySg, ec2.Port.allTraffic());
}
